File: ios/engage_flutter/Sources/engage_flutter/EngageCodec.swift
import Foundation
import UserNotifications
import EngageSDK

typealias FlutterMap = [String: Any]

enum EngageFlutterCodecError: Error {
  case invalidArgument(String)
}

extension Dictionary where Key == String, Value == Any {
  func string(_ key: String) throws -> String {
    guard let value = self[key] as? String else {
      throw EngageFlutterCodecError.invalidArgument("Missing string: \(key)")
    }
    return value
  }

  func bool(_ key: String) throws -> Bool {
    guard let value = self[key] as? Bool else {
      throw EngageFlutterCodecError.invalidArgument("Missing boolean: \(key)")
    }
    return value
  }

  func map(_ key: String) throws -> FlutterMap {
    guard let value = self[key] as? FlutterMap else {
      throw EngageFlutterCodecError.invalidArgument("Missing map: \(key)")
    }
    return value
  }

  func list(_ key: String) -> [Any] { self[key] as? [Any] ?? [] }
}

func engageConfig(_ value: FlutterMap) throws -> EngageConfig {
  let push = try value.map("push")
  let foreground = ForegroundPresentation(rawFlutter: try push.string("foregroundPresentation"))
  let ios = push["ios"] as? FlutterMap
  let categories = try Set((ios?.list("categories") ?? []).map(notificationCategory))
  let endpointValue = try value.string("endpoint")
  guard let endpoint = URL(string: endpointValue),
        let scheme = endpoint.scheme?.lowercased(),
        ["http", "https"].contains(scheme),
        endpoint.host != nil else {
    throw EngageFlutterCodecError.invalidArgument("Invalid HTTP(S) endpoint: \(endpointValue)")
  }
  return EngageConfig(
    appKey: try value.string("appKey"),
    endpoint: endpoint,
    push: PushConfig(
      foregroundPresentation: foreground,
      notificationCategories: categories
    )
  )
}

private func notificationCategory(_ value: Any) throws -> UNNotificationCategory {
  guard let category = value as? FlutterMap else {
    throw EngageFlutterCodecError.invalidArgument("Invalid iOS notification category")
  }
  let actions = try category.list("actions").map(notificationAction)
  return UNNotificationCategory(
    identifier: try category.string("key"),
    actions: actions,
    intentIdentifiers: [],
    hiddenPreviewsBodyPlaceholder: category["hiddenPreviewsBodyPlaceholder"] as? String,
    options: []
  )
}

private func notificationAction(_ value: Any) throws -> UNNotificationAction {
  guard let action = value as? FlutterMap else {
    throw EngageFlutterCodecError.invalidArgument("Invalid iOS notification action")
  }
  var options: UNNotificationActionOptions = []
  if (action["foreground"] as? Bool) == true { options.insert(.foreground) }
  if (action["destructive"] as? Bool) == true { options.insert(.destructive) }
  if (action["authenticationRequired"] as? Bool) == true {
    options.insert(.authenticationRequired)
  }
  return UNNotificationAction(
    identifier: try action.string("key"),
    title: try action.string("title"),
    options: options
  )
}

private extension ForegroundPresentation {
  init(rawFlutter: String) {
    self = rawFlutter == "SILENT" ? .silent : .show
  }
}

func jsonValue(_ value: Any?) throws -> JSONValue {
  switch value {
  case nil, is NSNull: return .null
  case let value as Bool: return .bool(value)
  case let value as Int: return .integer(Int64(value))
  case let value as Int64: return .integer(value)
  case let value as NSNumber:
    let double = value.doubleValue
    guard double.isFinite else {
      throw EngageFlutterCodecError.invalidArgument("JSON numbers must be finite")
    }
    return .number(double)
  case let value as String: return .string(value)
  case let value as [Any]: return .array(try value.map(jsonValue))
  case let value as FlutterMap:
    return .object(try value.mapValues(jsonValue))
  default:
    throw EngageFlutterCodecError.invalidArgument("Unsupported JSON value: \(type(of: value))")
  }
}

func flutterValue(_ value: JSONValue) -> Any {
  switch value {
  case .null: return NSNull()
  case let .bool(value): return value
  case let .integer(value): return value
  case let .number(value): return value
  case let .string(value): return value
  case let .array(value): return value.map(flutterValue)
  case let .object(value): return value.mapValues(flutterValue)
  }
}

private func flutterOptional<T>(_ value: T?) -> Any {
  guard let value else { return NSNull() }
  return value
}

func flutterPushStatus(_ status: PushStatus) -> FlutterMap {
  [
    "permission": status.permission.rawValue,
    "subscription": status.subscription.rawValue,
    "tokenRegistered": status.tokenRegistered,
  ]
}

func flutterPushEvent(_ event: PushEvent) -> FlutterMap {
  switch event {
  case let .received(deliveryId, messageId, data):
    return ["type": "RECEIVED", "deliveryId": deliveryId, "messageId": messageId, "data": data]
  case let .opened(deliveryId, messageId, deepLink, data):
    return [
      "type": "OPENED", "deliveryId": deliveryId, "messageId": messageId,
      "deepLink": flutterOptional(deepLink?.absoluteString), "data": data,
    ]
  case let .dismissed(deliveryId, messageId):
    return ["type": "DISMISSED", "deliveryId": deliveryId, "messageId": messageId]
  case let .actionSelected(deliveryId, messageId, actionKey, data):
    return [
      "type": "ACTION_SELECTED", "deliveryId": deliveryId, "messageId": messageId,
      "actionKey": actionKey, "data": data,
    ]
  case let .registrationFailed(message):
    return ["type": "REGISTRATION_FAILED", "message": message]
  }
}

func flutterInAppContent(_ content: InAppContent) -> FlutterMap {
  [
    "experienceId": content.experienceId,
    "messageId": content.messageId,
    "variantId": flutterOptional(content.variantId),
    "type": content.type.rawValue,
    "payload": content.payload.mapValues(flutterValue),
    "presentation": flutterPresentation(content.presentation),
  ]
}

private func flutterPresentation(_ presentation: PresentationSpec) -> FlutterMap {
  switch presentation {
  case let .overlay(value):
    return [
      "kind": "OVERLAY", "format": value.format.rawValue,
      "position": flutterOptional(value.position?.rawValue), "backdrop": value.backdrop.rawValue,
      "dismissal": value.dismissal.rawValue, "animation": value.animation.rawValue,
      "autoDismissAfterSeconds": flutterOptional(value.autoDismissAfterSeconds),
    ]
  case let .embedded(value):
    return [
      "kind": "EMBEDDED", "placementKey": value.placementKey,
      "emptyState": value.emptyState.rawValue,
    ]
  }
}

func flutterPreferenceCenter(_ center: PreferenceCenterSnapshot) -> FlutterMap {
  [
    "key": center.key,
    "displayName": center.displayName,
    "description": flutterOptional(center.description),
    "sections": center.sections.map { section in
      [
        "key": section.key,
        "title": flutterOptional(section.title),
        "description": flutterOptional(section.description),
        "subscriptions": section.subscriptions.map { subscription in
          [
            "key": subscription.key,
            "displayName": subscription.displayName,
            "description": flutterOptional(subscription.description),
            "profileChoices": flutterOptional(subscription.profileChoices?.reduce(into: FlutterMap()) {
              $0[$1.key.rawValue] = $1.value
            }),
            "installationChoice": flutterOptional(subscription.installationChoice),
          ]
        },
      ] as FlutterMap
    },
  ]
}

func flutterPagerState(_ state: InboxPagerState) -> FlutterMap {
  [
    "entries": state.entries.map { entry in
      [
        "id": entry.id.value,
        "key": entry.key,
        "payload": entry.payload.mapValues(flutterValue),
        "sentAt": iso8601(entry.sentAt),
        "expiresAt": flutterOptional(entry.expiresAt.map(iso8601)),
        "readAt": flutterOptional(entry.readAt.map(iso8601)),
      ] as FlutterMap
    },
    "isRefreshing": state.isRefreshing,
    "isLoadingMore": state.isLoadingMore,
    "hasMore": state.hasMore,
    "error": flutterOptional(state.error.map { error in
      [
        "code": inboxErrorCode(error.code),
        "message": error.message,
        "isRetryable": error.isRetryable,
      ] as FlutterMap
    }),
  ]
}

private func inboxErrorCode(_ code: InboxErrorCode) -> String {
  switch code {
  case .network: return "NETWORK"
  case .unauthorized: return "UNAUTHORIZED"
  case .generationChanged: return "GENERATION_CHANGED"
  case .server: return "SERVER"
  case .invalidResponse: return "INVALID_RESPONSE"
  case .localPersistence: return "LOCAL_PERSISTENCE"
  }
}

private func iso8601(_ date: Date) -> String {
  ISO8601DateFormatter().string(from: date)
}

File: android/src/main/kotlin/io/engage/engage_flutter/EngageCodec.kt
package io.engage.engage_flutter

import android.content.Context
import io.engage.sdk.AndroidPushAction
import io.engage.sdk.AndroidPushCategory
import io.engage.sdk.AndroidPushChannel
import io.engage.sdk.AndroidPushConfig
import io.engage.sdk.AndroidPushSound
import io.engage.sdk.EmbeddedPresentation
import io.engage.sdk.EngageConfig
import io.engage.sdk.EngageLogLevel
import io.engage.sdk.InAppContent
import io.engage.sdk.InboxEntry
import io.engage.sdk.InboxError
import io.engage.sdk.InboxPagerState
import io.engage.sdk.NotificationImportance
import io.engage.sdk.OverlayPresentation
import io.engage.sdk.PreferenceCenterSnapshot
import io.engage.sdk.PresentationSpec
import io.engage.sdk.PushConfig
import io.engage.sdk.PushEvent
import io.engage.sdk.PushStatus
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.doubleOrNull
import kotlinx.serialization.json.longOrNull
import java.net.URI

internal typealias FlutterMap = Map<String, Any?>

internal fun Any?.asMap(): FlutterMap = (this as? Map<*, *>)
    ?.entries
    ?.associate { (key, value) -> key.toString() to value }
    ?: error("Expected a map, got ${this?.javaClass?.name}")

internal fun Any?.asMapOrEmpty(): FlutterMap =
    if (this == null) emptyMap() else asMap()

internal fun Any?.asList(): List<Any?> = (this as? List<*>)?.toList().orEmpty()

internal inline fun <reified T : Enum<T>> enumValue(raw: Any?): T = enumValueOf(raw as String)

internal fun FlutterMap.string(key: String): String = requireNotNull(this[key] as? String) {
    "Missing string: $key"
}

internal fun FlutterMap.boolean(key: String): Boolean = requireNotNull(this[key] as? Boolean) {
    "Missing boolean: $key"
}

internal fun FlutterMap.toEngageConfig(context: Context): EngageConfig {
    val push = this["push"].asMap()
    return EngageConfig(
        appKey = string("appKey"),
        endpoint = URI.create(string("endpoint")),
        logLevel = when (val raw = this["logLevel"] as? String ?: "INFO") {
            "WARNING" -> EngageLogLevel.WARN
            else -> enumValue(raw)
        },
        push = PushConfig(
            foregroundPresentation = enumValue(push["foregroundPresentation"]),
            android = push["android"]?.asMap()?.toAndroidPushConfig(context),
        ),
    )
}

private fun FlutterMap.toAndroidPushConfig(context: Context): AndroidPushConfig {
    val resources = context.resources
    val packageName = context.packageName
    fun resource(name: String, type: String): Int {
        val normalized = name.substringAfterLast('/').substringAfterLast('.')
        val id = resources.getIdentifier(normalized, type, packageName)
        require(id != 0) { "Android resource $type/$normalized does not exist in $packageName" }
        return id
    }
    return AndroidPushConfig(
        smallIcon = resource(string("smallIconResource"), "drawable"),
        accentColor = (this["accentColorResource"] as? String)?.let { resource(it, "color") },
        defaultChannelKey = string("defaultChannelKey"),
        channels = this["channels"].asList().map { value ->
            val channel = value.asMap()
            val sound = channel["sound"].asMap()
            AndroidPushChannel(
                key = channel.string("key"),
                name = resource(channel.string("nameResource"), "string"),
                description = (channel["descriptionResource"] as? String)?.let { resource(it, "string") },
                importance = enumValue<NotificationImportance>(channel["importance"]),
                showBadge = channel.boolean("showBadge"),
                sound = when (sound.string("type")) {
                    "DEFAULT" -> AndroidPushSound.Default
                    "SILENT" -> AndroidPushSound.Silent
                    "RESOURCE" -> AndroidPushSound.Resource(resource(sound.string("rawResource"), "raw"))
                    else -> error("Unsupported Android push sound: ${sound["type"]}")
                },
            )
        },
        categories = this["categories"].asList().map { value ->
            val category = value.asMap()
            AndroidPushCategory(
                key = category.string("key"),
                actions = category["actions"].asList().map { actionValue ->
                    val action = actionValue.asMap()
                    AndroidPushAction(
                        key = action.string("key"),
                        title = resource(action.string("titleResource"), "string"),
                        opensApp = action.boolean("opensApp"),
                    )
                },
            )
        },
    )
}

internal fun Any?.toJsonElement(): JsonElement = when (this) {
    null -> JsonNull
    is JsonElement -> this
    is String -> JsonPrimitive(this)
    is Boolean -> JsonPrimitive(this)
    is Byte -> JsonPrimitive(toInt())
    is Short -> JsonPrimitive(toInt())
    is Int -> JsonPrimitive(this)
    is Long -> JsonPrimitive(this)
    is Float -> JsonPrimitive(toDouble())
    is Double -> JsonPrimitive(this)
    is Map<*, *> -> JsonObject(entries.associate { (key, value) -> key.toString() to value.toJsonElement() })
    is List<*> -> JsonArray(map { it.toJsonElement() })
    else -> error("Unsupported JSON value: ${javaClass.name}")
}

internal fun JsonElement.toFlutter(): Any? = when (this) {
    JsonNull -> null
    is JsonObject -> entries.associate { (key, value) -> key to value.toFlutter() }
    is JsonArray -> map(JsonElement::toFlutter)
    is JsonPrimitive -> when {
        isString -> content
        booleanOrNull != null -> booleanOrNull
        longOrNull != null -> longOrNull
        doubleOrNull != null -> doubleOrNull
        else -> contentOrNull
    }
}

internal fun PushStatus.toFlutter(): FlutterMap = mapOf(
    "permission" to permission.name,
    "subscription" to subscription.name,
    "tokenRegistered" to tokenRegistered,
)

internal fun PushEvent.toFlutter(): FlutterMap = when (this) {
    is PushEvent.Received -> mapOf(
        "type" to "RECEIVED",
        "deliveryId" to deliveryId,
        "messageId" to messageId,
        "data" to data,
    )
    is PushEvent.Opened -> mapOf(
        "type" to "OPENED",
        "deliveryId" to deliveryId,
        "messageId" to messageId,
        "deepLink" to deepLink,
        "data" to data,
    )
    is PushEvent.Dismissed -> mapOf(
        "type" to "DISMISSED",
        "deliveryId" to deliveryId,
        "messageId" to messageId,
    )
    is PushEvent.ActionSelected -> mapOf(
        "type" to "ACTION_SELECTED",
        "deliveryId" to deliveryId,
        "messageId" to messageId,
        "actionKey" to actionKey,
        "data" to data,
    )
}

internal fun InAppContent.toFlutter(): FlutterMap = mapOf(
    "experienceId" to experienceId,
    "messageId" to messageId,
    "variantId" to variantId,
    "type" to type.name,
    "payload" to payload.toFlutter(),
    "presentation" to presentation.toFlutter(),
)

private fun PresentationSpec.toFlutter(): FlutterMap = when (this) {
    is OverlayPresentation -> mapOf(
        "kind" to "OVERLAY",
        "format" to format.name,
        "position" to position?.name,
        "backdrop" to backdrop.name,
        "dismissal" to dismissal.name,
        "animation" to animation.name,
        "autoDismissAfterSeconds" to autoDismissAfterSeconds,
    )
    is EmbeddedPresentation -> mapOf(
        "kind" to "EMBEDDED",
        "placementKey" to placementKey,
        "emptyState" to emptyState.name,
    )
}

internal fun PreferenceCenterSnapshot.toFlutter(): FlutterMap = mapOf(
    "key" to key,
    "displayName" to displayName,
    "description" to description,
    "sections" to sections.map { section ->
        mapOf(
            "key" to section.key,
            "title" to section.title,
            "description" to section.description,
            "subscriptions" to section.subscriptions.map { subscription ->
                mapOf(
                    "key" to subscription.key,
                    "displayName" to subscription.displayName,
                    "description" to subscription.description,
                    "profileChoices" to subscription.profileChoices?.mapKeys { it.key.name },
                    "installationChoice" to subscription.installationChoice,
                )
            },
        )
    },
)

internal fun InboxPagerState.toFlutter(): FlutterMap = mapOf(
    "entries" to entries.map(InboxEntry::toFlutter),
    "isRefreshing" to isRefreshing,
    "isLoadingMore" to isLoadingMore,
    "hasMore" to hasMore,
    "error" to error?.toFlutter(),
)

internal fun InboxEntry.toFlutter(): FlutterMap = mapOf(
    "id" to id.value,
    "key" to key,
    "payload" to payload.toFlutter(),
    "sentAt" to sentAt.toString(),
    "expiresAt" to expiresAt?.toString(),
    "readAt" to readAt?.toString(),
)

private fun InboxError.toFlutter(): FlutterMap = mapOf(
    "code" to code.name,
    "message" to message,
    "isRetryable" to isRetryable,
)
